rusty-machine/src/linalg/matrix/iter.rs

```diff
@@ -459,4 +514,46 @@ mod tests {
 
     }
 
+    #[test]
+    #[allow(unused_variables)]
+    fn into_iter_compile() { 
+        let a = Matrix::new(3, 3, vec![2.0; 9]); 
+        let mut b = MatrixSlice::from_matrix(&a, [1, 1], 2, 2);
+    
+        for v in b { 
+        } 
+    
+        for v in &b { 
+        } 
+    
+        for v in &mut b { 
+        } 
+    } 
+    
+    #[test]
+    #[allow(unused_variables)]
+    fn into_iter_mut_compile() { 
+        let mut a = Matrix::new(3, 3, vec![2.0; 9]); 
+        
+        {
+            let b = MatrixSliceMut::from_matrix(&mut a, [1, 1], 2, 2);
+    
+            for v in b { 
+            } 
+        }
+    
+        {
+            let b = MatrixSliceMut::from_matrix(&mut a, [1, 1], 2, 2);
+    
+            for v in &b { 
+            } 
+        }
+    
+        {
+            let mut b = MatrixSliceMut::from_matrix(&mut a, [1, 1], 2, 2);
+    
+            for v in &mut b { 
```

Comment:
I think it's worth adding an

`*v = 1.0;`

here. This will add a further check to the compilation that the value returned is mutable. (And also the same for the `for v in b` loop.